parts/k8s/kubernetesmastergenerateetcdcertscript.sh

```diff
@@ -0,0 +1,101 @@
+ETCD_PEER_CRTS=("${ETCD_PEER_CRT_0:=/tmp/etcd-peer0.crt}" "${ETCD_PEER_CRT_1:=/tmp/etcd-peer1.crt}" "${ETCD_PEER_CRT_2:=/tmp/etcd-peer2.crt}" "${ETCD_PEER_CRT_3:=/tmp/etcd-peer3.crt}" "${ETCD_PEER_CRT_4:=/tmp/etcd-peer4.crt}")
+K8S_ETCD_PEER_CRT_FILEPATH="${K8S_ETCD_PEER_CRT_FILEPATH:=/etc/kubernetes/certs/etcd-peer${1}.crt}"
+K8S_ETCD_PEER_KEY_FILEPATH="${K8S_ETCD_PEER_KEY_FILEPATH:=/etc/kubernetes/certs/etcd-peer${1}.key}"
+
+# etcd tls client certs
+ETCD_CLIENT_KEY="${ETCD_CLIENT_KEY:=/tmp/etcd-client.key}"
+ETCD_CLIENT_CSR="${ETCD_CLIENT_CSR:=/tmp/etcd-client.csr}"
+ETCD_CLIENT_CRT="${ETCD_CLIENT_CRT:=/tmp/etcd-client.crt}"
+K8S_ETCD_CLIENT_CRT_FILEPATH="${K8S_ETCD_CLIENT_CRT_FILEPATH:=/etc/kubernetes/certs/etcd-client.crt}"
+K8S_ETCD_CLIENT_KEY_FILEPATH="${K8S_ETCD_CLIENT_KEY_FILEPATH:=/etc/kubernetes/certs/etcd-client.key}"
+
+# etcd files for writing certs to disk
+ETCD_REQUESTHEADER_CA="${ETCD_REQUESTHEADER_CA:=/etcdcerts/requestheader-etcd-ca-file}"
+ETCD_SERVER_CERT_FILE="${ETCD_SERVER_CERT_FILE:=/etcdcerts/etcd-server-cert-file}"
+ETCD_SERVER_KEY_FILE="${ETCD_SERVER_KEY_FILE:=/etcdcerts/etcd-server-key-file}"
+ETCD_CLIENT_CERT_FILE="${ETCD_CLIENT_CERT_FILE:=/etcdcerts/etcd-client-cert-file}"
+ETCD_CLIENT_KEY_FILE="${ETCD_CLIENT_KEY_FILE:=/etcdcerts/etcd-client-key-file}"
+ETCD_PEER_CERT_FILES=("${ETCD_PEER_CERT_FILE_0:=/etcdcerts/etcd-peer-cert-file-0}" "${ETCD_PEER_CERT_FILE_1:=/etcdcerts/etcd-peer-cert-file-1}" "${ETCD_PEER_CERT_FILE_2:=/etcdcerts/etcd-peer-cert-file-2}" "${ETCD_PEER_CERT_FILE_3:=/etcdcerts/etcd-peer-cert-file-3}" "${ETCD_PEER_CERT_FILE_4:=/etcdcerts/etcd-peer-cert-file-4}")
+ETCD_PEER_KEY_FILES=("${ETCD_PEER_KEY_FILE_0:=/etcdcerts/etcd-peer-key-file-0}" "${ETCD_PEER_KEY_FILE_1:=/etcdcerts/etcd-peer-key-file-1}" "${ETCD_PEER_KEY_FILE_2:=/etcdcerts/etcd-peer-key-file-2}" "${ETCD_PEER_KEY_FILE_3:=/etcdcerts/etcd-peer-key-file-3}" "${ETCD_PEER_KEY_FILE_4:=/etcdcerts/etcd-peer-key-file-4}")
+
+echo subjectAltName = IP:127.0.0.1 > extfile.cnf
+
+# generate root CA
+openssl genrsa -out $ETCD_CA_KEY 2048
+openssl req -new -x509 -days 1826 -key $ETCD_CA_KEY -out $ETCD_CA_CRT -subj '/CN=etcdCA'
+# generate new cert
+openssl genrsa -out $ETCD_SERVER_KEY 2048
+openssl req -new -key $ETCD_SERVER_KEY -out $ETCD_SERVER_CSR -subj '/CN=127.0.0.1/O=system:masters'
+openssl x509 -req -days 730 -in $ETCD_SERVER_CSR -CA $ETCD_CA_CRT -CAkey $ETCD_CA_KEY -set_serial 02 -out $ETCD_SERVER_CRT -extfile extfile.cnf
+openssl genrsa -out $ETCD_CLIENT_KEY 2048
+openssl req -new -key $ETCD_CLIENT_KEY -out $ETCD_CLIENT_CSR -subj '/CN=127.0.0.1/O=system:masters'
+openssl x509 -req -days 730 -in $ETCD_CLIENT_CSR -CA $ETCD_CA_CRT -CAkey $ETCD_CA_KEY -set_serial 02 -out $ETCD_CLIENT_CRT -extfile extfile.cnf
+for ((i = 0; i < ${#ETCD_PEER_KEYS[@]}; ++i)); do
+    openssl genrsa -out ${ETCD_PEER_KEYS[$i]} 2048
+    openssl req -new -key ${ETCD_PEER_KEYS[$i]} -out ${ETCD_PEER_CSRS[$i]} -subj '/CN=127.0.0.1/O=system:masters'
```

Comment:
This is to generate a certificate signing request for each etcd node in peer communication. The subject field should use the real IP for each peer node, or I think it will fail SSL hostname verification.

Btw, do we need the "/O" portion in the subject?